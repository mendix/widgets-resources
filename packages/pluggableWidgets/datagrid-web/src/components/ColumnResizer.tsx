import { createElement, ReactElement, useCallback, useEffect, useRef, useState, MouseEvent, TouchEvent } from "react";

export interface ColumnResizerProps {
    minWidth?: number;
    setColumnWidth: (width: number) => void;
    onResizeEnds?: () => void;
}

export function ColumnResizer({ minWidth = 50, setColumnWidth, onResizeEnds }: ColumnResizerProps): ReactElement {
    const [isResizing, setIsResizing] = useState(false);
    const [startPosition, setStartPosition] = useState(0);
    const [currentWidth, setCurrentWidth] = useState(0);
    const resizerReference = useRef<HTMLDivElement>(null);

    const onStartDrag = useCallback(
        (e: TouchEvent<HTMLDivElement> & MouseEvent<HTMLDivElement>): void => {
            const mouseX = e.touches ? e.touches[0].screenX : e.screenX;
            setStartPosition(mouseX);
            setIsResizing(true);
            if (resizerReference.current) {
                const column = resizerReference.current.parentElement as HTMLDivElement;
                setCurrentWidth(column.clientWidth);
            }
        },
        [resizerReference.current]
    );
    const onEndDrag = useCallback((): void => {
        if (!isResizing) {
            return;
        }
        setIsResizing(false);
        setCurrentWidth(0);
        onResizeEnds?.();
    }, [onResizeEnds, isResizing]);
    const onMouseMove = useCallback(
        (e: TouchEvent & MouseEvent & Event): void => {
            if (!isResizing) {
                return;
            }
            const mouseX = e.touches ? e.touches[0].screenX : e.screenX;

            if (currentWidth) {
                const moveDifference = startPosition - mouseX;
                let newWidth = currentWidth - moveDifference;
                if (newWidth < minWidth) {
                    newWidth = minWidth;
                }
                setColumnWidth(newWidth);
            }
        },
        [isResizing, currentWidth, startPosition]
    );

    useEffect(() => {
        document.addEventListener("mousemove", onMouseMove);
        document.addEventListener("mouseup", onEndDrag);

        document.addEventListener("touchmove", onMouseMove);
        document.addEventListener("touchend", onEndDrag);
        return () => {
            document.removeEventListener("mousemove", onMouseMove);
            document.removeEventListener("mouseup", onEndDrag);

            document.removeEventListener("touchmove", onMouseMove);
            document.removeEventListener("touchend", onEndDrag);
        };
    }, [onMouseMove, onEndDrag]);

    return (
        <div ref={resizerReference} className="column-resizer" onMouseDown={onStartDrag} onTouchStart={onStartDrag}>
            <div className="column-resizer-bar" />
        </div>
    );
}
