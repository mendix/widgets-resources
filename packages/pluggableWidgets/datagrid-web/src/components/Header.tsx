import { createElement, Dispatch, ReactElement, SetStateAction, DragEvent, DragEventHandler, useCallback } from "react";
import classNames from "classnames";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faLongArrowAltDown, faLongArrowAltUp, faArrowsAltV } from "@fortawesome/free-solid-svg-icons";
import { ColumnProperty } from "./Table";
import { ColumnResizerProps } from "./ColumnResizer";
import { SortingRule } from "../utils/settings";

export interface HeaderProps {
    className?: string;
    column: ColumnProperty;
    sortable: boolean;
    resizable: boolean;
    filterable: boolean;
    draggable: boolean;
    dragOver: string;
    hidable: boolean;
    isDragging?: boolean;
    preview?: boolean;
    resizer: ReactElement<ColumnResizerProps>;
    setColumnOrder: (updater: string[]) => void;
    setDragOver: Dispatch<SetStateAction<string>>;
    setIsDragging: Dispatch<SetStateAction<boolean>>;
    setSortBy: Dispatch<SetStateAction<SortingRule[]>>;
    sortBy: SortingRule[];
    visibleColumns: ColumnProperty[];
}

export function Header(props: HeaderProps): ReactElement {
    const canSort = props.sortable && props.column.canSort;
    const canDrag = props.draggable && (props.column.canDrag ?? false);
    const draggableProps = useDraggable(
        canDrag,
        props.visibleColumns,
        props.setColumnOrder,
        props.setDragOver,
        props.setIsDragging
    );

    const [sortProperties] = props.sortBy;
    const isSorted = sortProperties && sortProperties.id === props.column.id;
    const isSortedDesc = isSorted && sortProperties.desc;

    const sortIcon = canSort
        ? isSorted
            ? isSortedDesc
                ? faLongArrowAltDown
                : faLongArrowAltUp
            : faArrowsAltV
        : undefined;

    const caption = props.column.header;

    const onSortBy = (): void => {
        /**
         * Always analyse previous values to predict the next
         * 1 - !isSorted turns to asc
         * 2 - isSortedDesc === false && isSorted turns to desc
         * 3 - isSortedDesc === true && isSorted turns to unsorted
         * If multisort is allowed in the future this should be changed to append instead of just return a new array
         */
        if (!isSorted) {
            props.setSortBy([{ id: props.column.id, desc: false }]);
        } else if (isSorted && !isSortedDesc) {
            props.setSortBy([{ id: props.column.id, desc: true }]);
        } else {
            props.setSortBy([]);
        }
    };

    return (
        <div
            className={classNames("th", {
                "hidden-column-preview": props.preview && props.hidable && props.column.hidden
            })}
            role="columnheader"
            style={!props.sortable || !props.column.canSort ? { cursor: "unset" } : undefined}
            title={caption}
        >
            <div
                className={classNames(
                    "column-container",
                    canDrag && props.column.id === props.dragOver ? "dragging" : ""
                )}
                id={props.column.id}
                {...draggableProps}
            >
                <div
                    className={classNames("column-header", canSort ? "clickable" : "", props.className)}
                    id={props.column.id}
                    onClick={canSort ? onSortBy : undefined}
                    onKeyDown={
                        canSort
                            ? e => {
                                  if (e.key === "Enter" || e.key === " ") {
                                      e.preventDefault();
                                      onSortBy();
                                  }
                              }
                            : undefined
                    }
                    role={canSort ? "button" : undefined}
                    style={{ pointerEvents: props.isDragging ? "none" : undefined }}
                    tabIndex={canSort ? 0 : undefined}
                >
                    <span>{caption.length > 0 ? caption : "\u00a0"}</span>
                    {sortIcon && <FontAwesomeIcon icon={sortIcon} />}
                </div>
                {props.filterable && props.column.customFilter ? props.column.customFilter : null}
            </div>
            {props.resizable && props.column.canResize && props.resizer}
        </div>
    );
}

function useDraggable(
    columnsDraggable: boolean,
    visibleColumns: ColumnProperty[],
    setColumnOrder: (updater: ((columnOrder: string[]) => string[]) | string[]) => void,
    setDragOver: Dispatch<SetStateAction<string>>,
    setIsDragging: Dispatch<SetStateAction<boolean>>
): {
    draggable?: boolean;
    onDragStart?: DragEventHandler;
    onDragOver?: DragEventHandler;
    onDrop?: DragEventHandler;
    onDragEnter?: DragEventHandler;
    onDragEnd?: DragEventHandler;
} {
    const handleDragStart = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            setIsDragging(true);
            const { id } = e.target as HTMLDivElement;
            e.dataTransfer.setData("colDestination", id);
        },
        [setIsDragging]
    );

    const handleDragOver = useCallback((e: DragEvent<HTMLDivElement>): void => {
        e.preventDefault();
    }, []);

    const handleDragEnter = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            const { id } = e.target as HTMLDivElement;
            const colDestination = e.dataTransfer.getData("colDestination");
            if (id !== colDestination) {
                setDragOver(id);
            }
        },
        [setDragOver]
    );

    const handleDragEnd = useCallback((): void => {
        setIsDragging(false);
        setDragOver("");
    }, [setDragOver, setIsDragging]);

    const handleOnDrop = useCallback(
        (e: DragEvent<HTMLDivElement>): void => {
            handleDragEnd();
            const { id: colOrigin } = e.target as HTMLDivElement;
            const colDestination = e.dataTransfer.getData("colDestination");

            const toIndex = visibleColumns.findIndex(col => col.id === colOrigin);
            const fromIndex = visibleColumns.findIndex(col => col.id === colDestination);

            if (toIndex !== fromIndex) {
                const newOrder = [...visibleColumns.map(column => column.id)];
                newOrder.splice(fromIndex, 1);
                newOrder.splice(toIndex, 0, colDestination);
                setColumnOrder(newOrder);
            }
        },
        [handleDragEnd, setColumnOrder, visibleColumns]
    );

    return columnsDraggable
        ? {
              draggable: true,
              onDragStart: handleDragStart,
              onDragOver: handleDragOver,
              onDrop: handleOnDrop,
              onDragEnter: handleDragEnter,
              onDragEnd: handleDragEnd
          }
        : {};
}
