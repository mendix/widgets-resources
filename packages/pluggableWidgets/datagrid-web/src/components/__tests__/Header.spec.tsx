import { shallow } from "enzyme";
import { createElement } from "react";
import { Header, HeaderProps } from "../Header";
import { ColumnResizer } from "../ColumnResizer";

describe("Header", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<Header {...mockHeaderProps()} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when sortable", () => {
        const props = mockHeaderProps();
        props.column.canSort = true;
        props.sortable = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when resizable", () => {
        const props = mockHeaderProps();
        props.column.canResize = true;
        props.resizable = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when draggable", () => {
        const props = mockHeaderProps();
        props.column.canDrag = true;
        props.draggable = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when filterable with no custom filter", () => {
        const props = mockHeaderProps();
        props.filterable = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when filterable with custom filter", () => {
        const props = mockHeaderProps();
        props.column.customFilter = (
            <div>
                <label>Date picker filter</label>
                <input type="date" />
            </div>
        );
        props.filterable = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });

    it("calls setSortBy store function with correct parameters when sortable", () => {
        const column = {
            id: "sortable",
            header: "My sortable column",
            canSort: true
        } as any;
        const mockedFunction = jest.fn();
        const component = shallow(
            <Header {...mockHeaderProps()} column={column} sortable setSortBy={mockedFunction} />
        );

        const clickableRegion = component.find(".column-header");

        expect(clickableRegion).toHaveLength(1);

        clickableRegion.simulate("click");
        expect(mockedFunction).toBeCalledWith([{ id: "sortable", desc: false }]);
    });

    it("renders the structure correctly when filterable with custom classes", () => {
        const props = mockHeaderProps();
        props.filterable = true;

        const component = shallow(<Header {...props} className="my-custom-class" />);

        expect(component).toMatchSnapshot();
    });

    it("renders the structure correctly when is hidden and preview", () => {
        const props = mockHeaderProps();
        props.column.hidden = true;
        props.hidable = true;
        props.preview = true;

        const component = shallow(<Header {...props} />);

        expect(component).toMatchSnapshot();
    });
});

function mockHeaderProps(): HeaderProps {
    return {
        column: {
            header: "Test"
        } as any,
        draggable: false,
        dragOver: "",
        filterable: false,
        hidable: false,
        resizable: false,
        resizer: <ColumnResizer setColumnWidth={jest.fn()} />,
        sortable: false,
        setColumnOrder: jest.fn(),
        setDragOver: jest.fn(),
        visibleColumns: [],
        setSortBy: jest.fn(),
        setIsDragging: jest.fn(),
        sortBy: []
    };
}
