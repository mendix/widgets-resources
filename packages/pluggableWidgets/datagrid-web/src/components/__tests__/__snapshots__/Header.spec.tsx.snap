// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders the structure correctly 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when draggable 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
    draggable={true}
    onDragEnd={[Function]}
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDragStart={[Function]}
    onDrop={[Function]}
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with custom classes 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header my-custom-class"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with custom filter 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
    <div>
      <label>
        Date picker filter
      </label>
      <input
        type="date"
      />
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when filterable with no custom filter 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when is hidden and preview 1`] = `
<div
  className="th hidden-column-preview"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
</div>
`;

exports[`Header renders the structure correctly when resizable 1`] = `
<div
  className="th"
  role="columnheader"
  style={
    Object {
      "cursor": "unset",
    }
  }
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
    >
      <span>
        Test
      </span>
    </div>
  </div>
  <ColumnResizer
    setColumnWidth={[MockFunction]}
  />
</div>
`;

exports[`Header renders the structure correctly when sortable 1`] = `
<div
  className="th"
  role="columnheader"
  title="Test"
>
  <div
    className="column-container"
  >
    <div
      className="column-header clickable"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      style={
        Object {
          "pointerEvents": undefined,
        }
      }
      tabIndex={0}
    >
      <span>
        Test
      </span>
      <FontAwesomeIcon
        border={false}
        className=""
        fixedWidth={false}
        flip={null}
        icon={
          Object {
            "icon": Array [
              256,
              512,
              Array [],
              "f338",
              "M214.059 377.941H168V134.059h46.059c21.382 0 32.09-25.851 16.971-40.971L144.971 7.029c-9.373-9.373-24.568-9.373-33.941 0L24.971 93.088c-15.119 15.119-4.411 40.971 16.971 40.971H88v243.882H41.941c-21.382 0-32.09 25.851-16.971 40.971l86.059 86.059c9.373 9.373 24.568 9.373 33.941 0l86.059-86.059c15.12-15.119 4.412-40.971-16.97-40.971z",
            ],
            "iconName": "arrows-alt-v",
            "prefix": "fas",
          }
        }
        inverse={false}
        listItem={false}
        mask={null}
        pull={null}
        pulse={false}
        rotation={null}
        size={null}
        spin={false}
        swapOpacity={false}
        symbol={false}
        title=""
        transform={null}
      />
    </div>
  </div>
</div>
`;
